ceramic-kubo-rpc/src/dag.rs

```diff
@@ -0,0 +1,280 @@
+        let mut field = item.map_err(|e| Error::InternalError(e.into()))?;
+        if field.name() == "file" {
+            let mut input_bytes: Vec<u8> = Vec::new();
+            while let Some(chunk) = field.next().await {
+                input_bytes.extend(&chunk.map_err(|e| Error::InternalError(e.into()))?.to_vec())
+            }
+
+            let dag_data: Ipld = match query.input_codec.as_str() {
+                DAG_JSON => Ipld::decode(DagJsonCodec, &mut Cursor::new(&input_bytes))
+                    .map_err(|e| Error::BadRequest(e.into()))?,
+                _ => {
+                    return Err(Error::BadRequest(anyhow!(
+                        "unsupported input-codec \"{}\"",
+                        query.input_codec
+                    )));
+                }
+            };
+
+            let mut blob: Vec<u8> = Vec::with_capacity(input_bytes.len());
+            let codec: u64 = match query.store_codec.as_str() {
+                DAG_CBOR => {
+                    dag_data
+                        .encode(DagCborCodec, &mut blob)
+                        .map_err(|e| Error::InternalError(e.into()))?;
+                    DagCborCodec.into()
+                }
+
+                _ => {
+                    return Err(Error::BadRequest(anyhow!(
+                        "unsupported store-codec \"{}\"",
+                        query.store_codec
+                    )));
+                }
+            };
+
+            let store = data.api.try_store().map_err(|e| Error::InternalError(e))?;
+            let hash = Code::Sha2_256.digest(&blob);
+            let cid = Cid::new_v1(codec, hash);
+            store
+                .put(cid, blob.into(), vec![])
+                .await
+                .map_err(|e| Error::InternalError(e))?;
+            return Ok(HttpResponse::Ok().into());
+        }
+    }
+    Err(Error::BadRequest(anyhow!("missing multipart field 'file'")))
+}
+
+#[derive(Debug, Deserialize)]
+struct ResolveQuery {
+    arg: String,
+}
+
+#[tracing::instrument(skip(data))]
+async fn resolve<T>(
+    data: web::Data<AppState<T>>,
+    query: web::Query<ResolveQuery>,
+) -> Result<HttpResponse, Error>
+where
+    T: IrohClient,
+{
+    let path: IpfsPath = query.arg.parse().map_err(|e| Error::BadRequest(e))?;
+    let resolved_path = data
+        .api
+        .resolve(&path)
+        .await
+        .map_err(|e| Error::InternalError(e))?;
+
+    let cid = resolved_path
+        .iter()
+        .last()
+        .ok_or(Error::InternalError(anyhow!(
+            "resolved path should have at least one element"
+        )))?
+        .to_owned();
+    let resolved = ipld!({
+        "Cid": cid,
+        // TODO(nathanielc): What is this? Do we use it?
+        "RemPath": "",
+    });
+    let mut data = Vec::new();
+    resolved.encode(DagJsonCodec, &mut data).unwrap();
+    Ok(HttpResponse::Ok()
+        .content_type(ContentType::json())
+        .body(data))
+}
+
+#[cfg(test)]
```

Comment:
Note, these tests do not pass. I seem to have found a bug in actix-web  or am not using their testing API correctly. I have pinged on their discord hoping to get some help.

Reply:
I have since added tests for all implemented endpoints.